src/Utilities/Functional.hpp

```diff
@@ -0,0 +1,155 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cmath>
+#include <cstddef>
+#include <tuple>
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Higher order function objects similar to `std::plus`, etc.
+ */
+namespace funcl {
+/// \cond
+template <size_t Arity>
+struct Functional {
+  static constexpr size_t arity = Arity;
+
+ protected:
+  template <class C, size_t Offset, class... Ts, size_t... Is>
+  static constexpr decltype(auto) helper(
+      const std::tuple<Ts...>& t,
+      std::index_sequence<Is...> /*meta*/) noexcept {
+    return C{}(std::get<Offset + Is>(t)...);
+  }
+};
+
+struct Identity;
+/// \endcond
+
+#define MAKE_BINARY_OPERATOR(NAME, OPERATOR)                            \
+  template <class C0 = Identity, class C1 = C0>                         \
+  struct NAME : Functional<C0::arity + C1::arity> {                     \
+    using base = Functional<C0::arity + C1::arity>;                     \
+    template <class... Ts>                                              \
+    constexpr auto operator()(const Ts&... ts) noexcept {               \
+      return base::template helper<C0, 0>(                              \
```

Comment:
if accessing the helper is awkward, why not just make it a separate function? It doesn't need access to the class members. it could also be put in a detail namespace to hide it away.